File: Tests/TestProjects/Shared/Operation Tests/AKFMOscillatorTests.m
//
//  AKFMOscillatorTests.m
//  iOSObjectiveCAudioKit
//
//  Created by Aurelius Prochazka on 5/22/15.
//  Copyright (c) 2015 Aurelius Prochazka. All rights reserved.
//

#import "AKTestCase.h"

#define testDuration 10.0

@interface TestFMOscillatorInstrument : AKInstrument
@end

@implementation TestFMOscillatorInstrument

- (instancetype)init
{
    self = [super init];
    if (self) {


        AKLine *frequencyLine = [[AKLine alloc] initWithFirstPoint:akp(10)
                                                       secondPoint:akp(880)
                                             durationBetweenPoints:akp(testDuration)];
        AKLine *carrierMultiplierLine = [[AKLine alloc] initWithFirstPoint:akp(2)
                                                               secondPoint:akp(0)
                                                     durationBetweenPoints:akp(testDuration)];
        AKLine *modulatingMultiplierLine = [[AKLine alloc] initWithFirstPoint:akp(0)
                                                                  secondPoint:akp(2)
                                                        durationBetweenPoints:akp(testDuration)];
        AKLine *indexLine = [[AKLine alloc] initWithFirstPoint:akp(0)
                                                   secondPoint:akp(30)
                                         durationBetweenPoints:akp(testDuration)];
        // Instrument Definition
        AKFMOscillator *oscillator = [AKFMOscillator oscillator];
        oscillator.baseFrequency = frequencyLine;
        oscillator.carrierMultiplier = carrierMultiplierLine;
        oscillator.modulatingMultiplier = modulatingMultiplierLine;
        oscillator.modulationIndex = indexLine;

        [self setAudioOutput:oscillator];
    }
    return self;
}

@end

@interface AKFMOscillatorTests : AKTestCase
@end

@implementation AKFMOscillatorTests

- (void)testFMOscillator
{
    // Set up performance
    TestFMOscillatorInstrument *testInstrument = [[TestFMOscillatorInstrument alloc] init];
    [AKOrchestra addInstrument:testInstrument];
    [testInstrument playForDuration:testDuration];

    // Check output
    XCTAssertEqualObjects([self md5ForOutputWithDuration:testDuration], @"dfe4b8c87584f8847acc1352ba3b2bf2");
}

- (void)testPresetBuzzer
{
    AKInstrument *testInstrument = [AKInstrument instrument];
    [testInstrument setAudioOutput:[AKFMOscillator presetBuzzer]];
    [AKOrchestra addInstrument:testInstrument];
    [testInstrument play];

    // Check output
    XCTAssertEqualObjects([self md5ForOutputWithDuration:1.0], @"850cd857039adb870a83573a972ecd08");
}

- (void)testPresetFoghorn
{
    AKInstrument *testInstrument = [AKInstrument instrument];
    [testInstrument setAudioOutput:[AKFMOscillator presetFogHorn]];
    [AKOrchestra addInstrument:testInstrument];
    [testInstrument play];

    // Check output
    XCTAssertEqualObjects([self md5ForOutputWithDuration:1.0], @"214908a4ee92b0696692e6e841c916f6");
}

- (void)testPresetSpaceWobble
{
    AKInstrument *testInstrument = [AKInstrument instrument];
    [testInstrument setAudioOutput:[AKFMOscillator presetSpaceWobble]];
    [AKOrchestra addInstrument:testInstrument];
    [testInstrument play];

    // Check output
    XCTAssertEqualObjects([self md5ForOutputWithDuration:1.0], @"ff899014015ea961fa208d42cb0875e6");
}

- (void)testPresetSpiral
{
    AKInstrument *testInstrument = [AKInstrument instrument];
    [testInstrument setAudioOutput:[AKFMOscillator presetSpiral]];
    [AKOrchestra addInstrument:testInstrument];
    [testInstrument play];

    // Check output
    XCTAssertEqualObjects([self md5ForOutputWithDuration:1.0], @"0ae1436fd2f47c2ad487ef9bdfa55c26");
}

- (void)testPresetStunRay
{
    AKInstrument *testInstrument = [AKInstrument instrument];
    [testInstrument setAudioOutput:[AKFMOscillator presetStunRay]];
    [AKOrchestra addInstrument:testInstrument];
    [testInstrument play];

    // Check output
    XCTAssertEqualObjects([self md5ForOutputWithDuration:1.0], @"8386b8c1e289c702846b97cc9eccc641");
}

- (void)testPresetWobble
{
    AKInstrument *testInstrument = [AKInstrument instrument];
    [testInstrument setAudioOutput:[AKFMOscillator presetWobble]];
    [AKOrchestra addInstrument:testInstrument];
    [testInstrument play];

    // Check output
    XCTAssertEqualObjects([self md5ForOutputWithDuration:1.0], @"79b9806fbea0d735ee8ea95a8d875737");
}

@end

File: AudioKit/Common/Tests/AKFMOscillatorTests.swift
//
//  AKFMOscillatorTests.swift
//  AudioKit
//
//  Created by Aurelius Prochazka on 8/4/16.
//  Copyright © 2017 Aurelius Prochazka. All rights reserved.
//

import AudioKit
import XCTest

class AKFMOscillatorTests: AKTestCase {

    func testDefault() {
        output = AKFMOscillator()
        AKTestMD5("3fee66d7da663b3e68142de923962819")
    }

    func testParametersSetAfterInit() {
        let oscillator = AKFMOscillator(waveform: AKTable(.square))
        oscillator.baseFrequency = 1_234
        oscillator.carrierMultiplier = 1.234
        oscillator.modulatingMultiplier = 1.234
        oscillator.modulationIndex = 1.234
        oscillator.amplitude = 0.5
        output = oscillator
        AKTestMD5("8387b7242dbb91c0a1f397a9bb9f2b06")
    }

    func testParametersSetOnInit() {
        output = AKFMOscillator(waveform: AKTable(.square),
                                baseFrequency: 1_234,
                                carrierMultiplier: 1.234,
                                modulatingMultiplier: 1.234,
                                modulationIndex: 1.234,
                                amplitude: 0.5)
        AKTestMD5("8387b7242dbb91c0a1f397a9bb9f2b06")
    }

    func testPresetBuzzer() {
        let preset = AKFMOscillator()
        preset.presetBuzzer()
        output = preset
        AKTestMD5("03e2f9736e2511fe99997c65df486bbc")
    }

    func testPresetFogHorn() {
        let preset = AKFMOscillator()
        preset.presetFogHorn()
        output = preset
        AKTestMD5("e4e9fad0f2fc0c91b886583ae1e0faf4")
    }

    func testPresetSpiral() {
        let preset = AKFMOscillator()
        preset.presetSpiral()
        output = preset
        AKTestMD5("4cbeac11faec88c9816be8d872028657")
    }

    func testPresetStunRay() {
        let preset = AKFMOscillator()
        preset.presetStunRay()
        output = preset
        AKTestMD5("3434fc08a63bc6b0a8e52fbcc41e3866")
    }

    func testPresetWobble() {
        let preset = AKFMOscillator()
        preset.presetWobble()
        output = preset
        AKTestMD5("4450350ff43f5f1b258158f1ec7dbccc")
    }

    func testSquareWave() {
        output = AKFMOscillator(waveform: AKTable(.square, count: 4_096))
        AKTestMD5("521697a9ca4ef19632576bbc4f57e51f")
    }

}
